File: 0402-remove-k-digits/0402-remove-k-digits.java
import java.util.*;
import java.io.*;

class Solution {
    public String removeKdigits(String num, int k) {
        ArrayDeque<Character> ans = new ArrayDeque<Character>();

        for (char c : num.toCharArray()) {
            while (k > 0 && ans.size() > 0 && Integer.parseInt(String.valueOf(ans.peekLast())) > Integer.parseInt(String.valueOf(c))) {
                ans.removeLast();
                k--;
            }
            ans.addLast(c);
        }

        while (k > 0) {
            ans.removeLast();
            k--;
        }

        if (ans.size() == 0) {
            return "0";
        }
        
        while (ans.peekFirst() != '0') {
            ans.removeFirst();
        }
        return Arrays.toString(ans.toArray());
    }
}

File: 0402-remove-k-digits/0402-remove-k-digits.py
class Solution(object):
    def removeKdigits(self, num, k):
        ans = []
        for i in num:
            while k > 0 and ans and ans[-1] > i:
                ans.pop()
                k -= 1
            ans.append(i)

        while k > 0:
            ans.pop()
            k -= 1
        
        result = "".join(ans).lstrip("0")

        return result if result else "0"
        